src/Raven.Server/Documents/Handlers/BatchHandler.cs

```diff
@@ -285,16 +263,127 @@ await using (var writer = new AsyncBlittableJsonTextWriter(context, ResponseBody
+            
+                if (count.HasValue == false) // Failover and the ClusterTransactionCommand is already completed in the DocumentDatabase, so the task with the result is no longer exists, so we'll try to take the result from the local history log.
+                {
+                    using (ServerStore.Engine.ContextPool.AllocateOperationContext(out ClusterOperationContext ctx))
+                    using (ctx.OpenReadTransaction())
+                    {
+                        if (ServerStore.Engine.LogHistory.TryGetCommandByIndex(ctx, index, out var localCommand))
+                        {
+                            var clusterTxLocalResult = JsonConvert.DeserializeObject<ClusterTransactionResult>(localCommand["Result"].ToString());
+                            if (clusterTxLocalResult.Count.HasValue == false) //clusterTxLocalResult.Count == null and databaseCommandsCount>0 => clusterTxLocalResult.Count is 0 (DeserializeObject convert 0 to null)
+                                clusterTxLocalResult.Count = 0;
+            
+                            count = clusterTxLocalResult.Count.Value;
+                        }
+                        else
+                        {
+                            throw new InvalidOperationException("Cluster-transaction was succeeded, but Leader is outdated and its results are unaccessible.");
+                        }
+                    }
+                }
+            
+            }
+
+            return count;
+        }
+
         private static void ThrowNotSupportedType(string contentType)
         {
             throw new InvalidOperationException($"The requested Content type '{contentType}' is not supported. Use 'application/json' or 'multipart/mixed'.");
         }
 
+        private void ThrowClusterTransactionConcurrencyException(List<ClusterTransactionCommand.ClusterTransactionErrorInfo> errors)
+        {
+            HttpContext.Response.StatusCode = (int)HttpStatusCode.Conflict;
+            throw new ClusterTransactionConcurrencyException(
+                $"Failed to execute cluster transaction due to the following issues: {string.Join(Environment.NewLine, errors.Select(e => e.Message))}")
+            {
+                ConcurrencyViolations = errors.Select(e => e.Violation).ToArray()
+            };
+        }
+
+        private void GenerateDatabaseCommandsEvaluatedResults(List<ClusterTransactionDataCommand> databaseCommands, 
+            long index, long count, DateTime lastModified, bool? disableAtomicDocumentWrites,
+            DynamicJsonArray commandsResults)
+        {
+            if (count < 0)
+                throw new InvalidOperationException($"ClusterTransactionCommand result is invalid - count lower then 0 ({count}).");
+
+            foreach (var dataCmd in databaseCommands)
+            {
+
+                count++;
+                var cv = $"{ChangeVectorParser.RaftTag}:{count}-{Database.DatabaseGroupId}";
```

Comment:
We have this logic twice.
Maybe better to have one function and call it from both places

Reply:
Done